### PR TITLE
Avoid freeze in Windows API Level 0 and 1 console when UIA support is enabled

### LINKED ISSUE #14689: nvda logs a lot of '\r\n\r\n' if UIA support for Windows console is enabled
### Steps to reproduce:
1. In NVDA's Advanced Settings panel, make the following settings:
  - Windows Console support: combo box: UIA when available
2. Open cmd or git bash and press NVDA+Shift+M(Laptop);
### Actual behavior:
There are a lot of '\r\n' in the log.

### Expected behavior:
There won't be so many '\r\n' in the log
### NVDA logs, crash dumps and other attachments:
[log.txt](https://github.com/nvaccess/nvda/files/10866869/log.txt)

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
2023.1Beta2
#### Windows version:
Windows 10 22H2 (AMD64) build 19045.2604
#### Name and version of other software in use when reproducing the issue:
None
#### Other information about your system:
None
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
2022.4 is the same
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### PR DESCRIPTION
**Pull request by ABuffEr**
### Link to issue number:

Fixes #14689

### Summary of the issue:

Due to a bug in API Level 0 and 1 consoles (see [microsoft/terminal/#5243](https://github.com/microsoft/terminal/issues/5243)), sometimes NVDA freezes during dictionary processing of console text, containing a huge amount of blank lines.

### Description of user facing changes

None.

### Description of development approach

If current focus object has apiLevel 0 or 1, and difference between regular text and text with rightside blank lines stripped is greater than 100, we use cleaned text for dictionary processing.

### Testing strategy:

Manual, with #14689 str.

### Known issues with pull request:

None.

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [ ] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] Security precautions taken.

### CODE REVIEW

File: source/speechDictHandler/__init__.py
Comment (codeofdusk) [inline source/speechDictHandler/__init__.py:123]: This will fail if the config setting is `auto`. Can you check the API level of the console instead?

File: source/speechDictHandler/__init__.py
Comment (codeofdusk) [inline source/speechDictHandler/__init__.py:125]: Why 10000?

**Review by codeofdusk — CHANGES_REQUESTED** [review]

File: source/speechDictHandler/__init__.py
Comment (ABuffEr) [reply source/speechDictHandler/__init__.py:123]: But if config setting is auto UIA is not used at this API Level, and no freeze happen. Or do you mean that blank lines are present also in API Level>0 when UIA is automatically used?

Comment (codeofdusk) [reply source/speechDictHandler/__init__.py:123]: Blank lines are present before microsoft/terminal#6986 was merged, so API level <= `IMPROVED` I believe. I'd really prefer this be checked relative to the API level, not the config.

File: source/speechDictHandler/__init__.py
Comment (ABuffEr) [reply source/speechDictHandler/__init__.py:125]: I based on @CyrilleB79 's observations in #14689  (comment), and choose it to minimize intervention, but I can surely decrease it a lot.

File: source/speechDictHandler/__init__.py
Comment (ABuffEr) [reply source/speechDictHandler/__init__.py:123]: Ok, I have a better overview now. I'll try to check for apiLevel < 2.

File: source/speechDictHandler/__init__.py
Comment (ABuffEr) [reply source/speechDictHandler/__init__.py]: Making further tests, I observed a difference between 17998 and 17834, according to amount of text in console. Now that I use apiLevel on current focused object, however, I think we could reduce it to any value that guarantees no impact on performance and perceived spoken difference.

File: source/speechDictHandler/__init__.py
Comment (codeofdusk) [inline source/speechDictHandler/__init__.py:125]: Check for an instance of `WinConsoleUIA`.

**Review by codeofdusk — CHANGES_REQUESTED** [review]

File: source/speechDictHandler/__init__.py
Comment (ABuffEr) [reply source/speechDictHandler/__init__.py:125]: Like for WinConsoleAPILevel, apparently I should import WinConsoleUIA inside processText function to avoid crashes. And I'm afraid this could affect performance (but I accept denials, of course).

File: source/speechDictHandler/__init__.py
Comment (ABuffEr) [reply source/speechDictHandler/__init__.py]: @codeofdusk, do you have suggestions for this value at current stage? Thanks.

**Comment by codeofdusk** [comment]
Honestly, I'm not the biggest fan of this approach (or, for that matter, putting any more work into the API <2 console). There are changes coming down the pipeline that I'm not at liberty to discuss in detail (I would if I could), so I'd suggest holding off on any changes for now.

CC @DHowett.

File: source/speechDictHandler/__init__.py
Comment (seanbudd) [inline source/speechDictHandler/__init__.py:128]: It would be good to give this constant a name, and make it a bit smaller.
1000 trailing whitespace characters is already a lot. I'd even consider 100.

```suggestion
		IGNORE_TRAILING_WHITESPACE_LENGTH = 1000
		if len(text) - len(stripText) > IGNORE_TRAILING_WHITESPACE_LENGTH:
			text = stripText
```

**Comment by seanbudd** [comment]
>Honestly, I'm not the biggest fan of this approach (or, for that matter, putting any more work into the API <2 console). 

Shouldn't NVDA still support the API <2 console? Can you explain why not?
It seems like this small change will improve a freeze quite easily.

**Comment by ABuffEr** [comment]
@cary-rowen, can you test the last try-build when available?

File: source/speechDictHandler/__init__.py
Comment (codeofdusk) [inline source/speechDictHandler/__init__.py:129]: I'd really like a more precise check than `hasattr`, especially as NVDA can run external code in add-ons. Is there really no way of using `isinstance`?

File: source/speechDictHandler/__init__.py
Comment (seanbudd) [reply source/speechDictHandler/__init__.py:129]: Importing it inside the function should not effect performance

**Comment by ABuffEr** [comment]
Ok, I found a way that can accomplish @codeofdusk  requests.
I can import (into function) ConsoleUIATextInfo class, that shared by apiLevel 1 and 0 implementations, and then check istance of textInfo returned by review.getObjectPosition(globalVars.focusObject).
I could additionally check even apiLevel on focusObject, if you want, but it has not particularly sense at the moment

File: source/speechDictHandler/__init__.py
Comment (seanbudd) [inline source/speechDictHandler/__init__.py:119]: Can this remain private - either scoped in `processText` or prefixed with an underscore.

```suggestion
_IGNORE_TRAILING_WHITESPACE_LENGTH = 100
```

File: source/speechDictHandler/__init__.py
Comment (seanbudd) [inline source/speechDictHandler/__init__.py:125]: Why not use api.getReviewPosition?

```suggestion
	textInfo = api.getReviewPosition()
```

File: source/speechDictHandler/__init__.py
Comment (seanbudd) [inline source/speechDictHandler/__init__.py:131]:
```suggestion
	# late import to prevent circular dependency
```

File: source/speechDictHandler/__init__.py
Comment (seanbudd) [reply source/speechDictHandler/__init__.py:125]: This should at the very least use `api.getFocusObject`. `globalVars.focusObject` should not be used directly, as per the comment attached.

File: source/speechDictHandler/__init__.py
Comment (seanbudd) [inline source/speechDictHandler/__init__.py:136]: I really think it is easier to keep these together, mainly for readibility.

```suggestion
		IGNORE_TRAILING_WHITESPACE_LENGTH = 100
		if len(text) - len(stripText) > IGNORE_TRAILING_WHITESPACE_LENGTH:
```

**Review by seanbudd — APPROVED** [review]
Thanks @ABuffEr

**Comment by seanbudd** [comment]
@ABuffEr - you omitted the change log section from the PR template. Do you think this warrants a change log entry?
What do you think of this in bug fixes:
"When forcing UIA support with certain terminal and consoles, a bug is fixed which caused a freeze and the log file to be spammed. (#14689)"

**Review by codeofdusk — APPROVED** [review]
Approved pending my suggested comment clarifications (for consistency in style of references)

**Comment by ABuffEr** [comment]
> "When forcing UIA support with certain terminal and consoles, a bug is fixed which caused a freeze and the log file to be spammed. (#14689)"

Ok for me.

**Comment by ABuffEr** [comment]
Hi @seanbudd,
unfortunately this PR seems to cause troubles in some situations. I'm testing a new solution acting from _get_text directly in ConsoleUIATextInfo (new PR soon). Can you revert this? Thanks, and sorry for inconvenience.

**Comment by codeofdusk** [comment]
If you're able to fix the bug by patching `ConsoleUIATextInfo` directly, I too would feel much more comfortable with this.

**Comment by codeofdusk** [comment]
I've opened #14815.

**Pull request merged**
